### PR TITLE
CR and prefiltering in energy smoothing

### PR DESCRIPTION
**Pull request by ben-s-southworth**
Rewrote the CR implementation with a C backend. Removed some bugs from previous, deleted old python implementations. Added a pre-filtering option to the energy prolongation smoothing. CR has been tested a bit, pre-filtering less.

### CODE REVIEW

File: pyamg/aggregation/smooth.py [hunk 1]
```diff
@@ -904,7 +904,7 @@ def gmres_prolongation_smoothing(A, T, B, BtBinv, Sparsity_Pattern, maxiter,
 def energy_prolongation_smoother(A, T, Atilde, B, Bf, Cpt_params,
                                  krylov='cg', maxiter=4, tol=1e-8,
                                  degree=1, weighting='local',
-                                 Pfilter=None):
+                                 Pfilter=None, prefilter=None):
```
Comment (jbschroder) [inline pyamg/aggregation/smooth.py]: - How about changing "Pfilter" to "postfilter"?
- Unit tests

File: pyamg/aggregation/smooth.py [hunk 2]
```diff
@@ -1089,10 +1098,26 @@ def unpack_arg(v):
...
+                Sparsity_Pattern = truncate_rows(Sparsity_Pattern, kwargs['k'])
+                # Union two sparsity patterns
+                Sparsity_Pattern += Sparsity_theta
+            elif 'k' in kwargs:
+                Sparsity_Pattern = truncate_rows(Sparsity_Pattern, kwargs['k'])
+            elif 'theta' in kwargs:
+                Sparsity_Pattern = filter_matrix_rows(Sparsity_Pattern, kwargs['theta'])
+        elif fn is not None:
+            raise ValueError("Unrecognized prefilter option")
+
```
Comment (jbschroder) [inline pyamg/aggregation/smooth.py]: Looks good to me

File: pyamg/classical/cr.py [hunk 3]
```diff
@@ -3,290 +3,189 @@
 
 __docformat__ = "restructuredtext en"
 
+import pdb
+
```
Comment (jbschroder) [inline pyamg/classical/cr.py]: when done debugging, remove

File: pyamg/classical/cr.py [hunk 4]
```diff
@@ -3,290 +3,189 @@
...
-            # u[roots] = 1.0
-            # for depth in range(0,tdepth):
-            #     u = np.abs(S) * u
-            # (troots,tmp) = np.where(u>0)
-
-    return splitting.ravel()
-
-
-def _CRsweep(A, Findex, Cindex, nu, thetacr, method):
+def _CRsweep(A, B, Findex, Cindex, nu, thetacr, method):
```
Comment (jbschroder) [inline pyamg/classical/cr.py]: Could you add a comment describing the functionality

File: pyamg/classical/cr.py [hunk 5]
```diff
@@ -3,290 +3,189 @@
...
-        if rhok < 0.1 * thetacr:
-            return rhok, e
+    return rhok, e
 
 
-def CRalpha(A, method='habituated', nu=3, thetacr=0.7, thetacs=[0.3, 0.5],
-            maxiter=20):
-    """
+def CR(A, method='habituated', B=None, nu=3, thetacr=0.5,
+        thetacs='auto', maxiter=20, verbose=False):
```
Comment (jbschroder) [inline pyamg/classical/cr.py]: Maybe include 2 small unit tests that compare the CF splitting to a stored CF splitting.

File: pyamg/aggregation/smooth.py (hunk 1 above)
Comment (ben-s-southworth) [inline pyamg/aggregation/smooth.py]: I think we should remove "row-wise." In that case does it make sense for prefilter and postfilter be dictionaries with possible arguments 'k' and 'theta'?

File: pyamg/aggregation/smooth.py [hunk 6]
```diff
@@ -904,7 +904,7 @@ def gmres_prolongation_smoothing(A, T, B, BtBinv, Sparsity_Pattern, maxiter,
 def energy_prolongation_smoother(A, T, Atilde, B, Bf, Cpt_params,
                                  krylov='cg', maxiter=4, tol=1e-8,
                                  degree=1, weighting='local',
-                                 Pfilter=None):
+                                 prefilter=None, postfilter=None):
     """Minimize the energy of the coarse basis functions (columns of T).  Both
```
Comment (jbschroder) [inline pyamg/aggregation/smooth.py]: Default {}

File: pyamg/aggregation/smooth.py [hunk 7]
```diff
@@ -1166,6 +1190,7 @@ def unpack_arg(v):
                                          krylov=krylov, maxiter=1,
                                          tol=1e-8, degree=0,
                                          weighting=weighting,
-                                         Pfilter=('secondpass', {}))
+                                         prefilter=prefilter,
```
Comment (jbschroder) [inline pyamg/aggregation/smooth.py]: prefilter should be {}

**Pull request merged**